src/ragas/testset/persona.py

```diff
@@ -0,0 +1,150 @@
+    summaries: t.List[str]
+
+
+class Persona(BaseModel):
+    name: str
+    role_description: str
+
+
+class PersonasList(BaseModel):
+    personas: t.List[Persona]
+
+    def __getitem__(self, key: str) -> Persona:
+        for persona in self.personas:
+            if persona.name == key:
+                return persona
+        raise KeyError(f"No persona found with name '{key}'")
+
+
+# Define the prompt class
+class PersonaGenerationPrompt(PydanticPrompt[SummaryInput, PersonasList]):
+    instruction: str = (
+        "Using the provided summaries, generate one persona for each summary who might "
+        "interact with the content. For each persona, include a unique name "
+        "and a brief role description of who they are."
+    )
+    input_model: t.Type[SummaryInput] = SummaryInput
+    output_model: t.Type[PersonasList] = PersonasList
+    examples: t.List[t.Tuple[SummaryInput, PersonasList]] = [
+        (
+            SummaryInput(
+                summaries=[
+                    "Guide to Digital Marketing explains strategies for engaging audiences across various online platforms.",
+                    "Data Privacy Essentials discusses principles for safeguarding user data and complying with privacy regulations.",
+                    "Introduction to Project Management covers key methods for planning, executing, and monitoring projects.",
+                ]
+            ),
+            PersonasList(
+                personas=[
+                    Persona(
+                        name="Digital Marketing Specialist",
+                        role_description="Focuses on engaging audiences and growing the brand online.",
+                    ),
+                    Persona(
+                        name="Data Privacy Officer",
+                        role_description="Ensures the organization's compliance with data protection laws.",
+                    ),
+                    Persona(
+                        name="Project Manager",
+                        role_description="Oversees project timelines and ensures tasks are completed efficiently.",
+                    ),
+                ]
+            ),
+        )
+    ]
+
+
+@dataclass
+class PersonaGenerator:
+
+    llm: BaseRagasLLM
+    num_personas: int = 5
+    prompt: PydanticPrompt = PersonaGenerationPrompt()
+    filter_nodes: t.Callable[[Node], bool] = field(
+        default_factory=lambda: default_filter
+    )
+
+    def __post_init__(self):
+
+        try:
+            from sklearn.cluster import KMeans
+            from sklearn.metrics import pairwise_distances
+        except ImportError:
+            raise ImportError(
+                "PersonaGenerator requires the 'scikit-learn' package to be installed. "
+                "You can install it with 'pip install scikit-learn'."
+            )
+
+        self.pairwise_distances = pairwise_distances
+        self.kmeans = KMeans(n_clusters=self.num_personas, random_state=42)
```

Comment:
is it possible to generate this for ourselfs - this is a fairly popular interface so easily generateable

Reply:
Do you mean, is it possible to generate/manually write persona? Yes. It just needs to be in Persona Format

Reply:
no, is it important to add the sklearn dependency here? we can generate the code for the functionality here

Reply:
scikit-learn may be an overkill, I had the same doubt. Now I have reduced it to a much simpler algorithm w/o adding any extra dependency.